better performance in accessing children by tag name

banjun opened the pull request:
I hope fast Swift XML parser more fast! (comparable to Objective-C ones!)

Finding children with `func cXMLNode(tag:...)` takes extra cost of Swift.String generation from C String, and thus higher level String comparison `String.compare(...)` consume much cpu time (profiled by Instruments).

`func cXMLNode` takes libxml types. On the other hand, string literals can be converted `UnsafePointer<CChar>` (a.k.a. C String, `UnsafePointer<xmlChar>`, ...) via `Swift.StaticString`. Using StaticString, we can compare libxml tag name to the string on the libxml level `xmlStrcasecmp` without copying memory.

with this PR, `element.children(staticTag: "key")` to get children in better performance 🏎 

I have tried overloading `func children` for both String and StaticString, but string literals seems to fall to String always.
this cause `func firstChild` and `func children` to be duplicated into String version and StaticString versions.

cezheng commented on Sources/Element.swift:
```diff
@@ -119,13 +119,33 @@ open class XMLElement: XMLNode {
+
+  - returns: The child element.
+  */
+  open func firstChild(staticTag tag: StaticString, inNamespace ns: StaticString? = nil) -> XMLElement? {
```
I'm not too fond of adding yet another method just for accepting StaticString parameters. But since String and StaticString conform to some common protocols, maybe we can think of the possibility of accepting parameters conforming to one of these protocols, and use different implementations for String & StaticString. 

https://developer.apple.com/reference/swift/string#relationships
https://developer.apple.com/reference/swift/staticstring#relationships

cezheng commented on Sources/Element.swift:
```diff
@@ -134,7 +154,23 @@ open class XMLElement: XMLNode {
+
+  - returns: The children elements.
+  */
+  open func children(staticTag tag: StaticString, inNamespace ns: StaticString? = nil) -> [XMLElement] {
```
Same here

banjun commented:
thanks. I'll try that way!!

banjun commented:
* Added `XMLCharsComparable` protocol conformed exactly by `String` and `StaticString`.
* Made `cXMLNode(_:matchesTag:inNamespace:)` be just replacing the original compare with `XMLCharsComparable.caseInsensitivelyEqual`.
* For `firstChild and children` funcs, changed arg types to `XMLCharsComparable`.

Overload for `String` and `StaticString` is to be resolved the other way.

For `func f(tag: XMLCharsComparable)`, arg in `f(tag: "tagname")` is resolved as `String`. not as `StaticString`.
My last commit includes `children(staticTag: StaticString,...)` to avoid overloading mechanism and to use explicitly typed `StaticString`.
Another solution is adding type annotation on the caller side such as `children(tag: "tagname" as StaticString)`.

Anyway, caller side have to opt-in this by using either `children(staticTag: "tagname")` or `children(tag: "tagname" as StaticString)`.
I think it's OK to omit `func children(staticTag: StaticString)` explicitly typed version, although code completion won't suggest the omitted version. If so, caller side must opt-in with `"tagname" as StaticString` (maybe this kind of type annotations can be done with private extension on caller side).

cezheng commented:
um, so there will be no way to make a string literal be implicitly taken as a StaticString, right? @banjun

banjun commented:
Yes, I don't know much about Swift overloading priorities. In my several tries, strings literals are always falls to String, not overloaded StaticString version, without any type annotations or casts.

I've checked the swift docs today and found that choosing overloading uses some score based priorities. Each literal types has a default type. infer types as non-default type earns some negative priority. I guess String is the default type for string literals and thus StaticString always lose for overloading.

https://github.com/apple/swift/blob/master/docs/TypeChecker.rst#comparing-solutions

In another doc implies StringLiteralType is the default type for string literals.

https://github.com/apple/swift/blob/master/docs/Literals.rst

The doc says its module-scoped, but actually `fileprivate typealias` also works.
For example, the below patch **inverts** type inference priority among String and StaticString and that process with StaticString version of compare.
Even single `let` requires the opposite type annotation to be used as `String`. (see `let path: String = "/spec/header/title"`)

Have you ever used the `typealias StringLiteralType`?
Even with this solution, **the preference of StaticString than String** is to be specified by caller-side of Fuzi.
I cannot find a solution to prioritize StaticString by default and in safe, when the users use funcs in Fuzi.

```diff
diff --git a/Tests/XMLTests.swift b/Tests/XMLTests.swift
index 5532620..98aaee2 100644
--- a/Tests/XMLTests.swift
+++ b/Tests/XMLTests.swift
@@ -22,7 +22,11 @@
 import XCTest
 import Fuzi
 
+
+fileprivate typealias StringLiteralType = StaticString
+
 class XMLTests: XCTestCase {
+
   var document: Fuzi.XMLDocument!
   override func setUp() {
     super.setUp()
@@ -56,7 +60,7 @@ class XMLTests: XCTestCase {
   }
   
   func testXPath() {
-    let path = "/spec/header/title"
+    let path: String = "/spec/header/title"
     let elts = document.xpath(path)
     var counter = 0
     for elt in elts {
```

cezheng commented:
@banjun I see, thanks for the work:) Just add tests for the APIs you added then I think this PR is good to merge

banjun commented:
@cezheng Just added a test for that.

cezheng commented:
Thanks 🍻

banjun commented:
@cezheng any plan to release (to pod trunk)?
